tensorflow/contrib/opt/python/training/model_average_optimizer.py

```diff
@@ -0,0 +1,581 @@
+  , gradients will be averaged each step, and then applying them to the
+  variables in one shot, after which replicas can fetch the new variables and
+  continue. In a model average training environment, model variables will be
+  averaged (or with momentum) every 'ma_intervals' steps, and then fetch the new
+  variables and continue training in local worker. In the interval between two
+  "average operation", there are no data transfer at all, which can accerlate
+  training.
+
+  The following accumulators/queue are created:
+  <empty line>
+  * N `model-variable accumulators`, one per variable for train model. local
+  variables are pushed to them and the chief worker will wait until enough
+  variables are collected and then average them. The accumulator will drop all
+  stale variables (more details in the accumulator op).
+  * 1 `token` queue where the optimizer pushes the new global_step value after
+    all variables are updated.
+
+  The following local variable is created:
+  * `sync_rep_local_step`, one per replica. Compared against the global_step in
+    each accumulator to check for staleness of the variables.
+
+  The optimizer adds nodes to the graph to collect local variables and pause
+  the trainers until variables are updated.
+  For the Parameter Server job:
+  <empty line>
+  1. An accumulator is created for each variable, and each replica pushes the
+     local variables into the accumulators.
+  2. Each accumulator averages once enough variables (replicas_to_aggregate)
+     have been accumulated.
+  3. apply the averaged variables to global variables.
+  4. Only after all variables have been updated, increment the global step.
+  5. Only after step 4, pushes `global_step` in the `token_queue`, once for
+     each worker replica. The workers can now fetch the global step, use it to
+     update its local_step variable and start the next batch.
+
+  For the replicas:
+  <empty line>
+  1. Start a training block: fetch variables, finish "ma_intervals" steps
+     training.
+  2. Once current training block has been finished, push local variables into
+     accumulators. Each accumulator will check the staleness and drop the
+     stale ones.
+  3. After pushing all the variables, dequeue an updated value of global_step
+     from the token queue and record that step to its local_step variable. Note
+     that this is effectively a barrier.
+  4. fetch new variables, Start the next block.
+
+  ### Usage
+
+  ```python
+  # Create any optimizer to update the variables, say a simple SGD:
+  opt = GradientDescentOptimizer(learning_rate=0.1)
+
+  # Create a ModelAverageOptimizer to update the global variables:
+  # Note that if you want to have 2 backup replicas, you can change
+  # total_num_replicas=52 and make sure this number matches how many physical
+  # replicas you started in your job.
+  ma = tf.contrib.model_average.ModelAverageOptimizer(replicas_to_aggregate=50,
```

Comment:
We don't wrapper ModelAverageOptimizer as an actual TensorFlow Optimizer because we think it's unnatural to implement compute_gradients and apply_gradients Functions for model average.